numpy/core/src/umath/ufunc_object.c

```diff
@@ -1777,22 +1781,30 @@ execute_fancy_ufunc_loop(PyUFuncObject *ufunc,
 
     needs_api = NpyIter_IterationNeedsAPI(iter);
 
-    /* Copy any allocated outputs */
+    /* Call the __array_prepare__ functions where necessary */
     op_it = NpyIter_GetOperandArray(iter);
     for (i = nin; i < nop; ++i) {
-        if (op[i] == NULL) {
-            op[i] = op_it[i];
-            Py_INCREF(op[i]);
+        PyArrayObject *op_tmp;
+
+        /* Prepare_ufunc_output may decref & replace pointer */
+        op_tmp = op_it[i];
+        Py_INCREF(op_tmp);
+
+        if (prepare_ufunc_output(ufunc, &op_tmp,
+                                 arr_prep[i], arr_prep_args, i) < 0) {
+            NpyIter_Deallocate(iter);
+            return -1;
         }
-    }
 
-    /* Call the __array_prepare__ functions where necessary */
-    for (i = 0; i < nout; ++i) {
-        if (prepare_ufunc_output(ufunc, &op[nin+i],
-                            arr_prep[i], arr_prep_args, i) < 0) {
+        if (PyArray_BYTES(op_tmp) != PyArray_BYTES(op_it[i])) {
+            PyErr_SetString(PyExc_ValueError,
+                        "The __array_prepare__ functions modified the data "
+                        "pointer addresses in an invalid fashion");
+            Py_DECREF(op_tmp);
             NpyIter_Deallocate(iter);
             return -1;
         }
+        Py_DECREF(op_tmp);
```

Comment:
Now that I look at this again, I guess `op_tmp` should be placed back into `op` instead of being discarded.